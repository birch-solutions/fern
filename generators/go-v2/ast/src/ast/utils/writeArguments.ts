import { Argument, Arguments, isNamedArgument } from "@fern-api/base-generator";
import { Writer } from "../core/Writer";
import { TypeInstantiation } from "../TypeInstantiation";

export function writeArguments({ writer, arguments_ }: { writer: Writer; arguments_: Arguments }): void {
    const filteredArguments = filterNopArguments(arguments_);

    if (filteredArguments.length === 0) {
        writer.write("()");
        return;
    }
    writer.writeLine("(");
    writer.indent();
    for (const argument of filteredArguments) {
        writeArgument({ writer, argument });
        writer.writeLine(",");
    }
    writer.dedent();
    writer.write(")");
}

function writeArgument({ writer, argument }: { writer: Writer; argument: Argument }): void {
    if (isNamedArgument(argument)) {
        argument.assignment.write(writer);
    } else {
        argument.write(writer);
    }
}

function filterNopArguments(arguments_: Argument[]): Argument[] {
    return arguments_.filter(
        (argument) => !(argument instanceof TypeInstantiation && TypeInstantiation.isNop(argument))
    );
}
