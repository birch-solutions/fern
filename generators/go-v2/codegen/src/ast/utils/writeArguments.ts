import { Argument, Arguments, isNamedArgument } from "@fern-api/generator-commons";
import { Writer } from "../core/Writer";

export function writeArguments({ writer, arguments_ }: { writer: Writer; arguments_: Arguments }): void {
    if (arguments_.length === 0) {
        writer.write("()");
        return;
    }
    writer.writeLine("(");
    writer.indent();
    for (const argument of arguments_) {
        writeArgument({ writer, argument });
        writer.writeLine(",");
    }
    writer.dedent();
    writer.write(")");
}

function writeArgument({ writer, argument }: { writer: Writer; argument: Argument }): void {
    if (isNamedArgument(argument)) {
        argument.assignment.write(writer);
    } else {
        argument.write(writer);
    }
}
