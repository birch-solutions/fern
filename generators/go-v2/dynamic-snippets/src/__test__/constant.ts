import path from "path";
import { AbsoluteFilePath } from "@fern-api/fs-utils";

export const TEST_DEFINITIONS_DIRECTORY = AbsoluteFilePath.of(
    path.join(__dirname, "../../../../../packages/cli/dynamic-snippets/src/__test__/test-definitions")
);
