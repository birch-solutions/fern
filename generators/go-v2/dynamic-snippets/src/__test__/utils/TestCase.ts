import { dynamic as DynamicSnippets } from "@fern-fern/ir-sdk/api";

export interface TestCase {
    description: string;
    giveRequest: DynamicSnippets.EndpointSnippetRequest;
}
