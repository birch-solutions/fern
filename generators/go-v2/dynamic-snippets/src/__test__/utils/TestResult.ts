export class TestResult {
    public snippets: string[] = [];

    public addSnippet(snippet: string): void {
        this.snippets.push(snippet);
    }

    public toString(): string {
        if (this.snippets.length === 0) {
            return "<none>";
        }
        let s = "";
        this.snippets.forEach((snippet, idx) => {
            if (idx > 0) {
                s += "\n------------------------\n\n";
            }
            s += snippet;
        });
        return s;
    }
}
