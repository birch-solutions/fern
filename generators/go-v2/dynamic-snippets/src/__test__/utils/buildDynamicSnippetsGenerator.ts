import { AbsoluteFilePath, RelativeFilePath } from "@fern-api/fs-utils";
import { DynamicSnippetsGenerator } from "../../DynamicSnippetsGenerator";
import { FernGeneratorExec } from "@fern-api/generator-commons";
import { readFileSync } from "fs";
import { dynamic as DynamicSnippets } from "@fern-fern/ir-sdk/serialization";

export function buildDynamicSnippetsGenerator({
    irFilepath,
    config
}: {
    irFilepath: AbsoluteFilePath;
    config: FernGeneratorExec.GeneratorConfig;
}) {
    const content = readFileSync(irFilepath, "utf-8");
    const ir = DynamicSnippets.DynamicIntermediateRepresentation.parseOrThrow(JSON.parse(content));
    return new DynamicSnippetsGenerator({ ir, config });
}
