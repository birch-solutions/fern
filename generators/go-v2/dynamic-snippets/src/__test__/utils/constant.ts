import path from "path";
import { AbsoluteFilePath } from "@fern-api/fs-utils";

export const DYNAMIC_IR_TEST_DEFINITIONS_DIRECTORY = AbsoluteFilePath.of(
    path.join(
        __dirname,
        "../../../../../../packages/cli/generation/ir-generator/src/dynamic-snippets/__test__/test-definitions"
    )
);
