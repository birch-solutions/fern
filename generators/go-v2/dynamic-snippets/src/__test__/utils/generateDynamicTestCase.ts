import { Audiences } from "@fern-api/configuration";
import { createMockTaskContext } from "@fern-api/task-context";
import { AbstractAPIWorkspace } from "@fern-api/workspace-loader";
import { generateIntermediateRepresentation } from "@fern-api/ir-generator";
import { convertIrToDynamicSnippetsIr } from "@fern-api/dynamic-snippets";
import { IntermediateRepresentation, dynamic as DynamicSnippets } from "@fern-fern/ir-sdk/api";

export type DynamicTestCase = {
    ir: IntermediateRepresentation;
    dynamic: DynamicSnippets.DynamicIntermediateRepresentation;
};

export async function generateDynamicTestCase({
    workspace,
    audiences
}: {
    workspace: AbstractAPIWorkspace<unknown>;
    audiences: Audiences;
}): Promise<DynamicTestCase> {
    const context = createMockTaskContext();
    const fernWorkspace = await workspace.toFernWorkspace({
        context
    });
    const intermediateRepresentation = await generateIntermediateRepresentation({
        workspace: fernWorkspace,
        generationLanguage: undefined,
        audiences,
        keywords: undefined,
        smartCasing: true,
        disableExamples: false,
        readme: undefined,
        version: undefined,
        packageName: undefined,
        context
    });
    return {
        ir: intermediateRepresentation,
        dynamic: await convertIrToDynamicSnippetsIr(intermediateRepresentation)
    };
}
