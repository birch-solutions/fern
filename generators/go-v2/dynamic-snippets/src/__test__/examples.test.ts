import { buildDynamicSnippetsGenerator } from "./utils/buildDynamicSnippetsGenerator";
import { DYNAMIC_IR_TEST_DEFINITIONS_DIRECTORY } from "./utils/constant";
import { buildGeneratorConfig } from "./utils/buildGeneratorConfig";
import { AuthValues } from "@fern-fern/ir-sdk/api/resources/dynamic";
import { AbsoluteFilePath } from "@fern-api/path-utils";
import { TestCase } from "./utils/TestCase";

describe("examples", () => {
    const testCases: TestCase[] = [
        {
            description: "GET /metadata (simple)",
            giveRequest: {
                endpoint: {
                    method: "GET",
                    path: "/metadata"
                },
                auth: AuthValues.bearer({
                    token: "<YOUR_API_KEY>"
                }),
                pathParameters: undefined,
                queryParameters: {
                    shallow: false,
                    tag: "development"
                },
                headers: {
                    "X-API-Version": "0.0.1"
                },
                requestBody: undefined
            }
        },
        {
            description: "GET /metadata (allow-multiple)",
            giveRequest: {
                endpoint: {
                    method: "GET",
                    path: "/metadata"
                },
                auth: AuthValues.bearer({
                    token: "<YOUR_API_KEY>"
                }),
                pathParameters: undefined,
                queryParameters: {
                    shallow: false,
                    tag: ["development", "public"]
                },
                headers: {
                    "X-API-Version": "0.0.1"
                },
                requestBody: undefined
            }
        },
        {
            description: "POST /movie (simple)",
            giveRequest: {
                endpoint: {
                    method: "POST",
                    path: "/movie"
                },
                auth: AuthValues.bearer({
                    token: "<YOUR_API_KEY>"
                }),
                pathParameters: undefined,
                queryParameters: undefined,
                headers: undefined,
                requestBody: {
                    id: "movie-c06a4ad7",
                    prequel: "movie-cv9b914f",
                    title: "The Boy and the Heron",
                    from: "Hayao Miyazaki",
                    rating: 8.0,
                    type: "movie",
                    tag: "development",
                    metadata: {
                        actors: ["Christian Bale", "Florence Pugh", "Willem Dafoe"],
                        releaseDate: "2023-12-08",
                        ratings: {
                            rottenTomatoes: 97,
                            imdb: 7.6
                        }
                    },
                    revenue: 1000000
                }
            }
        },
        {
            description: "POST /big-entity (simple)",
            giveRequest: {
                endpoint: {
                    method: "POST",
                    path: "/big-entity"
                },
                auth: AuthValues.bearer({
                    token: "<YOUR_API_KEY>"
                }),
                pathParameters: undefined,
                queryParameters: undefined,
                headers: undefined,
                requestBody: {
                    castMember: {
                        id: "john.doe",
                        name: "John Doe"
                    }
                }
            }
        }
    ];
    const generator = buildDynamicSnippetsGenerator({
        irFilepath: AbsoluteFilePath.of(`${DYNAMIC_IR_TEST_DEFINITIONS_DIRECTORY}/examples.json`),
        config: buildGeneratorConfig()
    });
    test.each(testCases)("$description", async ({ giveRequest }) => {
        const response = await generator.generate(giveRequest);
        expect(response.snippet).toMatchSnapshot();
    });
});

describe("examples (errors)", () => {
    it("invalid request body", async () => {
        const generator = buildDynamicSnippetsGenerator({
        irFilepath: AbsoluteFilePath.of(`${DYNAMIC_IR_TEST_DEFINITIONS_DIRECTORY}/examples.json`),
            config: buildGeneratorConfig()
        });
        const response = await generator.generate({
            endpoint: {
                method: "POST",
                path: "/movie"
            },
            auth: AuthValues.bearer({
                token: "<YOUR_API_KEY>"
            }),
            pathParameters: undefined,
            queryParameters: undefined,
            headers: undefined,
            requestBody: {
                id: "movie-c06a4ad7",
                prequel: "movie-cv9b914f",
                title: 42, // invalid
                from: "Hayao Miyazaki",
                rating: 8.0,
                type: "movie",
                tag: "development",
                metadata: {
                    actors: ["Christian Bale", "Florence Pugh", "Willem Dafoe"],
                    releaseDate: "2023-12-08",
                    ratings: {
                        rottenTomatoes: 97,
                        imdb: 7.6
                    }
                },
                revenue: 1000000
            }
        });
        expect(response.errors).toMatchSnapshot();
    });
});
