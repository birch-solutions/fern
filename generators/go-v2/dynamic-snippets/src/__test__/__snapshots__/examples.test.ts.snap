// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`examples (errors) > invalid request body 1`] = `
[
  {
    "message": "Expected string but got number",
    "path": [
      "requestBody",
      "title",
    ],
    "severity": "CRITICAL",
  },
]
`;

exports[`examples > 'GET /metadata (allow-multiple)' 1`] = `
"package example

import (
	context "context"
	acme "github.com/acme/acme-go"
	client "github.com/acme/acme-go/client"
	option "github.com/acme/acme-go/option"
)

func do() {
	client := client.NewClient(
		option.WithToken(
			"<YOUR_API_KEY>",
		),
	)
	client.Service.GetMetadata(
		context.TODO(),
		&acme.GetMetadataRequest{
			Shallow: acme.Bool(
				false,
			),
			Tag: []*string{
				acme.String(
					"development",
				),
				acme.String(
					"public",
				),
			},
			XAPIVersion: "0.0.1",
		},
	)
}
"
`;

exports[`examples > 'GET /metadata (simple)' 1`] = `
"package example

import (
	context "context"
	acme "github.com/acme/acme-go"
	client "github.com/acme/acme-go/client"
	option "github.com/acme/acme-go/option"
)

func do() {
	client := client.NewClient(
		option.WithToken(
			"<YOUR_API_KEY>",
		),
	)
	client.Service.GetMetadata(
		context.TODO(),
		&acme.GetMetadataRequest{
			Shallow: acme.Bool(
				false,
			),
			Tag: []*string{
				acme.String(
					"development",
				),
			},
			XAPIVersion: "0.0.1",
		},
	)
}
"
`;

exports[`examples > 'POST /big-entity (simple)' 1`] = `
"package example

import (
	context "context"
	acme "github.com/acme/acme-go"
	client "github.com/acme/acme-go/client"
	option "github.com/acme/acme-go/option"
)

func do() {
	client := client.NewClient(
		option.WithToken(
			"<YOUR_API_KEY>",
		),
	)
	client.Service.CreateBigEntity(
		context.TODO(),
		&acme.BigEntity{
			CastMember: &acme.CastMember{
				Actor: &acme.Actor{
					ID:   "john.doe",
					Name: "John Doe",
				},
			},
		},
	)
}
"
`;

exports[`examples > 'POST /movie (simple)' 1`] = `
"package example

import (
	context "context"
	acme "github.com/acme/acme-go"
	client "github.com/acme/acme-go/client"
	option "github.com/acme/acme-go/option"
)

func do() {
	client := client.NewClient(
		option.WithToken(
			"<YOUR_API_KEY>",
		),
	)
	client.Service.CreateMovie(
		context.TODO(),
		&acme.Movie{
			ID: "movie-c06a4ad7",
			Prequel: acme.String(
				"movie-cv9b914f",
			),
			Title:  "The Boy and the Heron",
			From:   "Hayao Miyazaki",
			Rating: 8,
			Tag:    "development",
			Metadata: map[string]interface{}{
				"actors": []interface{}{
					"Christian Bale",
					"Florence Pugh",
					"Willem Dafoe",
				},
				"releaseDate": "2023-12-08",
				"ratings": map[string]interface{}{
					"rottenTomatoes": 97,
					"imdb":           7.6,
				},
			},
			Revenue: 1000000,
		},
	)
}
"
`;
